### PR TITLE
Add support for the scanning API

### PR DESCRIPTION
**Pull request by meskio**
I have created a new class `SdScanningClient` that exposes most of the secure scanning API. I thought it makes sense to be it's own class, as `SdSecureClient` is already pretty big and there are things like policies that exists both in the scanning part and in the secure part. If we want everything in `SdSecureClient` I can explore mixins or some inheritage tricks to have all the methods in one while the code is split in several files or just have one big class.

~~I have imported some files from anchore-cli into sdcclient/_anchore to bring all the API calls. At some point it might make sense to refactor all that out into the `SdScanningClient`, if you consider that a blocker to merge this pull-req I'm happy to work on that.~~ (I've cleaned it up, there is no anchore code anymore)

I have refactor SdMonitorClient and SdSecureClient in it's own files and clean up some pep8 errors, the _client.py file was getting too big.

### CODE REVIEW

File: sdcclient/_scanning.py [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+            url += "?history=true"
+        else:
+            url += "?history=false"
+        if detail:
+            url += "&detail=true"
+        else:
+            url += "&detail=false"
+        url += "&tag=" + str(thetag)
+        if policy:
+            url += "&policyId=" + str(policy)
```
Comment (figarocorso) [inline sdcclient/_scanning.py]: Just proposing a similar approach than before so we can read entire URL:
```python
url = "{base_url}/api/scanning/v1/anchore/images/{image_digest}/check?history={history}&detail={detail}&tag={tag}{policy}"
url = url.format(base_url=self.url,
                 image_digest=image_digest,
                 history=str(bool(show_history)).lower(),
                 detail=str(bool(detail)).lower(),
                 tag="&tag=%s" % thetag,
                 policy = "&policyId=%s" % policy if policy else '')
```

File: sdcclient/_scanning.py [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+
+        if registry_type and registry_type not in registry_types:
+            return [False, "input registry type not supported (supported registry_types: " + str(registry_types)]
+
+        # try to detect awsecr registry of form <accid>.dkr.ecr.<region>.amazonaws.com
+        if not registry_type:
+            if re.match("[0-9]+\\.dkr\\.ecr\\..*\\.amazonaws\\.com", registry):
+                registry_type = "awsecr"
+            else:
+                registry_type = "docker_v2"
```
Comment (figarocorso) [inline sdcclient/_scanning.py]: I would refactor this so no comments are needed (or they are documenting a method, not a code block):
```python
if not registry_type:
    registry_type = self._get_registry_type(registry)
```

File: sdcclient/_scanning.py [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+
+        # do some input string checking
+        if re.match(".*\\/.*", registry):
```
Comment (figarocorso) [inline sdcclient/_scanning.py]: We don't need a comment if we refactor this regex:
`if not self._registry_string_is_valid(registry):`

Moreover, we are using this code four times in the code. So this refactor is totally worthy.

File: sdcclient/_scanning.py [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+        itype = None
+
+        if re.match("^sha256:[0-9a-fA-F]{64}", input_string):
+            itype = 'imageDigest'
+        elif re.match("[0-9a-fA-F]{64}", input_string):
+            itype = 'imageid'
+        else:
+            itype = 'tag'
+
+        return(itype)
```
Comment (figarocorso) [inline sdcclient/_scanning.py]: We don't need the `(`, `)` here. But existing code has them all around, so :see_no_evil: 

We need a refactor there.

File: sdcclient/_scanning.py [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+        try:
+            digest = unquote_plus(str(input_string))
+            patt = re.match(".*(sha256:.*)", digest)
+            if patt:
+                return("imageDigest", input_string, input_string)
+            patt = re.match(".*(local:.*)", digest)
+            if patt:
+                return("imageDigest", input_string, input_string)
+        except Exception:
+            pass
```
Comment (figarocorso) [inline sdcclient/_scanning.py]: We can save a few lines here:
```python
try:
    sha256_patt = re.match(".*(sha256:.*)", digest)
    local_patt = re.match(".*(local:.*)", digest)
    if sha256_patt or local_patt:
        return("imageDigest", input_string, input_string)
except: Exception:
    pass
```

We can reduce try/except code block by refactoring:
```python
def _get_digest_pattern(self, input, local=False):
    try:
        return re.match(".*({}:.*)".format('local' if local else 'sha256'), input)
    except Exception:
        return False
```

File: sdcclient/_scanning.py [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+                    for image_detail in image_record['image_detail']:
+                        if input_string == image_detail['imageId']:
+                            ret_type = "imageid"
+                            break
+                except Exception:
+                    pass
+            else:
+                pass
+        except Exception:
+            urldigest = None
```
Comment (figarocorso) [inline sdcclient/_scanning.py]: What exception are we capturing here? Dict keys not found?
We can then simplify code:
```python
ret_type = "tag"
urldigest = None
ok, ret = self.get_image(input_string)
if ok:
    image_record = ret[0]
    urldigest = image_record.get('imageDigest', None)
    for image_detail in image_record.get('image_detail', []):
        if input_string == image_detail.get('imageId', ''):
            ret_type = "imageid"
            break
return(ret_type, input_string, urldigest)
```

File: sdcclient/_scanning.py [hunk 7]
```diff
@@ -428,6 +388,120 @@ def get_registry(self, registry):
...
+            return [False, self.lasterr]
+
+        return [True, res.json()]
+
+    def list_policy_bundles(self, detail=False):
+        url = self.url + "/api/scanning/v1/anchore/policies"
+        if detail:
+            url = url + "?detail=True"
+        else:
+            url = url + "?detail=False"
```
Comment (figarocorso) [inline sdcclient/_scanning.py]:
```python
url = self.url + "/api/scanning/v1/anchore/policies?detail=" + str(bool(detail))
```

File: sdcclient/_scanning.py [hunk 8]
```diff
@@ -558,6 +666,41 @@ def list_subscription(self):
...
+            'skipPolicyEvaluation': skip_policy_evaluation
+        }
+        if start_time or end_time:
+            containers['time'] = {}
+            if start_time:
+                containers['time']['from'] = int(start_time * 100000)
+            else:
+                containers['time']['from'] = 0
+            if not end_time:
+                end_time = time.time()
```
Comment (figarocorso) [inline sdcclient/_scanning.py]: Saving some lines:
```python
containers['time'] = {}
containers['time']['from'] = int(start_time * 100000) if start_time else 0
end_time = end_time if end_time else time.time()
```

File: sdcclient/_scanning.py [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+
+        **Success Return Value**
+            A JSON object representing the image that was imported.
+        '''
+        url = self.url + "/api/scanning/v1/anchore/imageimport"
+        res = requests.post(url, data=json.dumps(image_data), headers=self.hdrs, verify=self.ssl_verify)
+        if not self._checkResponse(res):
+            return [False, self.lasterr]
+
+        return [True, res.json()]
```
[the remark this answers is not in the document]
Comment (meskio) [reply sdcclient/_scanning.py]: Sounds good, I'll come with another PR for that as soon as we get this one merged.

File: sdcclient/_scanning.py (hunk 6 above)
Comment (meskio) [reply sdcclient/_scanning.py]: I'm not even sure what exception is. Most of this code is kind of copy-paste from anchore-cli, I should had cleaned it up more.

**Comment by omer-sd** [comment]
Travis is failing

**Comment by davideschiera** [comment]
@omer-sd the failure might not be related to the changes in the PR. I'll try to manually restart the build.

**Comment by davideschiera** [comment]
We have intermittent failures on the Travis build that are definitely unrelated to these changes. Last build succeeded.

**Comment by meskio** [comment]
The travis CI error says `PYTHON_SDC_TEST_ACCESS_KEY: unbound variable`, it looks like the API somehow didn't get set on this run.

@davideschiera any luck restarting the build manually?

**Comment by davideschiera** [comment]
@meskio The restarting part was easy, but unfortunately the build is still failing, and it's a configuration issue, nothing related to your changes. I started taking a look at the Travis configuration, but didn't find anything useful yet.

If @omer-sd is ok with this, an idea could be merging the PR anyway so that your work goes mainstream (the build should work as expected on dev/master branches), and in the meantime keep troubleshooting the build issue. @omer-sd please let me know. Thanks!

**Comment by omer-sd** [comment]
This is approved. We can merge the code. We will handle the Travis failure internally.

**Pull request merged**